### ISSUE TITLE
ES5 + downlevelIteration on spread operator disables strictNullChecks 

### ISSUE THREAD

**Issue by mprobst** [microsoft/TypeScript#29347]
**TypeScript Version:** 3.3.0-dev.20190110

**Search Terms:**

**Code**

```ts
declare const maybeArr: number[]|undefined;  // declare is load bearing, otherwise compiler seems to infer tighter type
const nums: number[] = [];
nums.push(...maybeArr);  // expected: undefined is not an array type, has no [Symbol.iterator]

// Actual practical instance where this came up:
const someMap = new Map<string, number[]>();
nums.push(...someMap.get('x'));  // expected: undefined is not an array type, has no [Symbol.iterator]
```

**Expected behavior:**

This should give a compile error with strictNullChecks, regardless of any other TypeScript compiler flags.

**Actual behavior:**

```sh
$ tsc --strictNullChecks --lib es2015 --target es5 --downlevelIteration test.ts
[ no error ]

$ tsc --strictNullChecks --lib es2015 --target es6 test.ts
test.ts:3:14 - error TS2488: Type 'undefined' must have a '[Symbol.iterator]()' method that returns an iterator.

3 nums.push(...maybeArr);
               ~~~~~~~~

test.ts:6:14 - error TS2488: Type 'undefined' must have a '[Symbol.iterator]()' method that returns an iterator.

6 nums.push(...someMap.get('x'));
               ~~~~~~~~~~~~~~~~
```

**Playground Link:** 
Link below does not show the problem in ES5/downlevelIteration mode.
http://www.typescriptlang.org/play/#src=declare%20const%20maybeArr%3A%20number%5B%5D%7Cundefined%3B%20%20%2F%2F%20declare%20is%20load%20bearing%2C%20otherwise%20compiler%20seems%20to%20infer%20tighter%20type%0Aconst%20nums%3A%20number%5B%5D%20%3D%20%5B%5D%3B%0Anums.push(...maybeArr)%3B%20%20%2F%2F%20expected%3A%20undefined%20is%20not%20an%20array%20type%2C%20has%20no%20%5BSymbol.iterator%5D%0A%0A%2F%2F%20Actual%20practical%20instance%20where%20this%20came%20up%3A%0Aconst%20someMap%20%3D%20new%20Map%3Cstring%2C%20number%5B%5D%3E()%3B%0Anums.push(...someMap.get('x'))%3B%20%20%2F%2F%20expected%3A%20undefined%20is%20not%20an%20array%20type%2C%20has%20no%20%5BSymbol.iterator%5D

**Related Issues:**

#17195 seems similar but was closed.

**Comment by a-tarasyuk** [comment]
When TS checks spread expression, it goes though the following chain `checkSpreadExpression -> checkIteratedTypeOrElementType -> getIteratedTypeOrElementType`. Inside the last method, there is the condition which allows report about error only if  `uplevelIteration ? errorNode : undefined`. As a result, TS does not report about errors if `--downlevelIteration` enabled and uses `--lib es2015` and `---target es5`.  I think it was done for some reason. @weswigham could you help with the understanding of this issue?
```typescript

const uplevelIteration = languageVersion >= ScriptTarget.ES2015;
const downlevelIteration = !uplevelIteration && compilerOptions.downlevelIteration;

// Get the iterated type of an `Iterable<T>` or `IterableIterator<T>` only in ES2015
// or higher, when inside of an async generator or for-await-if, or when
// downlevelIteration is requested.
if (uplevelIteration || downlevelIteration || allowAsyncIterables) {
    // We only report errors for an invalid iterable type in ES2015 or higher.
    const iteratedType = getIteratedTypeOfIterable(inputType, uplevelIteration ? errorNode : undefined, allowAsyncIterables, /*allowSyncIterables*/ true, checkAssignability);
    if (iteratedType || uplevelIteration) {
        return iteratedType;
    }
}
```

**Comment by RyanCavanaugh** [comment]
This is fixed now, TS reports
```
Type 'number[] | undefined' must have a '[Symbol.iterator]()' method that returns an iterator
```
